src/main/java/org/truffleruby/core/kernel/KernelNodes.java

```diff
@@ -1642,6 +1645,13 @@ private boolean respondToMissing(VirtualFrame frame, Object object, RubySymbol n
             return booleanCastNode.executeToBoolean(
                     respondToMissingNode.call(object, "respond_to_missing?", name, includeProtectedAndPrivate));
         }
+
+        private void useCallerRefinements(VirtualFrame frame) {
+            if (frame != null) {
+                DeclarationContext context = RubyArguments.getDeclarationContext(readCallerFrame.execute(frame));
```

Comment:
This is somewhat problematic, because getting the caller frame is fairly expensive, unless the caller is compiled together with `respond_to?`. And I expect there are many callers of `respond_to?`, so it will cause a lot of method to need to materialize their frame, which is bad for performance.

I think what we should do to address this concern is to AST-inline `respond_to?`.
That way, we don't need to read the caller frame, because we'd execute in the caller frame.
It's a bit similar to `Kernel#binding`: https://github.com/oracle/truffleruby/blob/5c94a6a3a6b24b77090a1a91f9352351ffe4a264/src/main/java/org/truffleruby/core/inlined/InlinedBindingNode.java

We'll still need this code in the case e.g. an alias of `respond_to?` is used, but in practice it should hopefully be almost never used.

Reply:
In order to AST-inline it, it seems best to extract a separate node with all the logic, except for `useCallerRefinements` which would then be only for this `@CoreMethod` node when it's not AST-inlined.

Reply:
The reason why choose using `ReadCallerFrameNode` is:

https://github.com/oracle/truffleruby/blob/e6b8e72b03f4579a9cfa2e566958543f10123180/src/main/java/org/truffleruby/core/kernel/KernelNodes.java#L1529-L1540

So I thought `public_send` and `respond_to?` methods of the same order of importance. Or I miss something?

Reply:
Indeed, `__send__/send` uses `ReadCallerFrameNode` too: https://github.com/oracle/truffleruby/blob/e6b8e72b03f4579a9cfa2e566958543f10123180/src/main/java/org/truffleruby/core/basicobject/BasicObjectNodes.java#L585-L605

Probably we should AST-inline those as well.
We had a recent deoptimization issue with `ReadCallerFrameNode` and in general trying to avoid `ReadCallerFrameNode` is best when not necessary.
So for this PR, let's keep the ReadCallerFrameNode approach, it should be good enough for now.

Reply:
I can take care of this in the next PR if you like...probably with your little help :)

Reply:
It seems OK for now, we'll try to fix that deoptimization issue with `ReadCallerFrameNode` first.

Reply:
Another way to fix this would be to have ReadCallerDeclarationContextNode or ReadCallerArgumentsNode, that's a lot cheaper to pass if not compiled & inlined.
We could also mark these few methods that always want the DeclarationContext of the caller and handle them in DispatchNode/CallInternalMethodNode.